packages/hub/tasks/check-reward-roots.ts

```diff
@@ -0,0 +1,57 @@
+import { inject } from '@cardstack/di';
+import WorkerClient from '../services/worker-client';
+import { RewardPrograms, RewardRoots } from '../services/subgraph';
+import { ProcessRewardRootPayload } from '../tasks/process-reward-root';
+import Logger from '@cardstack/logger';
+
+const MAX_INDEX_SIZE_PROGRAM = 1000;
+
+let log = Logger('task:check-reward-roots');
+export default class CheckRewardRoots {
+  subgraph = inject('subgraph');
+  databaseManager = inject('database-manager', { as: 'databaseManager' });
+  workerClient: WorkerClient = inject('worker-client', { as: 'workerClient' });
+  async perform(max_index_size_per_program = MAX_INDEX_SIZE_PROGRAM) {
+    try {
+      let db = await this.databaseManager.getClient();
+      const r1: RewardPrograms = await this.subgraph.getRewardPrograms();
+      const rewardProgramIds = r1?.data?.rewardPrograms.map((o) => o.id);
+      for (const rewardProgramId of rewardProgramIds) {
+        const LAST_INDEXED_BLOCK_QUERY = `SELECT COALESCE(MAX(block_number),0) as last_indexed_block_number FROM reward_root_index WHERE reward_program_id='${rewardProgramId}';`;
+        const {
+          rows: [{ last_indexed_block_number }],
+        } = await db.query(LAST_INDEXED_BLOCK_QUERY);
+        const r2: RewardRoots = await this.subgraph.getRewardRoots(
+          rewardProgramId,
+          last_indexed_block_number,
+          max_index_size_per_program
+        );
+        let payloads: ProcessRewardRootPayload[] = r2?.data?.merkleRootSubmissions.map((root) => {
+          return {
+            blockNumber: root.blockNumber,
+            s3FileInfo: {
+              paymentCycle: root.paymentCycle,
+              rewardProgramId: root.rewardProgram.id,
+            },
+          };
+        });
+        for (const payload of payloads) {
+          // Process 1 parquet (rewardProgramId, paymentCycle) file at one time
+          await this.workerClient.addJob('process-reward-root', payload, {
```

Comment:
Can you describe what the expected and maximum range of `rewardProgramIds` and `payloads` is at every run?

Reply:
Also, how long does `process-reward-root` task usually take?

Reply:
For now, 

production has 2 reward programs

every 10 minutes, max 1000 jobs will be executed per reward program. So at a maximum of 2000 jobs in total

using my local computer, the first run 1000+23 jobs will be indexed within 2 minutes of the worker running. In total, 67 proofs are indexed. So pretty sparse. In fact, this also assumes that I do not even check for expiry, if I check for expiry its probably 0 proofs.  

Most of the 1023 records are just storing `reward_root_index` which just keeps the indexer knwoing which reward_root is submitted. 

The speed of process_reward_root will depend on how many rows of rewards are being given. For example,

paymentCycle: 23979900 has 45 proofs (this was staking). The indexing is almost instant.

Reply:
Sounds like this won't bring our job processing to its knees for extended periods of time. Few more questions:

1. The first run currently takes a minute and then next runs before a new cycle won't have much work to do? How long does a cycle last?
2. The first run will blast the workers with work and it should be done in let's say minute like you benchmarked. I suppose this means any new work tasks that will enter the queue at this time will have to wait until a bunch reward tasks will be done? I think this can potentially be problematic - we have several tasks which we want to be started almost immediately, for example tasks of the "waiter" type, e.g `scheduled-payment-on-chain-creation-waiter`.  These tasks are coupled with feedback to user actions in the interface and we want to avoid having them enqueued for too long - they have to start as soon as possible. So my concern is that in the reward task task spike these other tasks will be delayed for a while. For this reason, have you considered introducing a separate queue for tasks that aren't critical to be performed immediately after enqueuing (I suppose reward tasks can be labelled as such)?

Reply:
1. The first run currently takes a minute and then next runs before a new cycle won't have much work to do? How long does a cycle last?

It depends on the rule. But in general we use a rule of thumb for staking. Staking rule runs every month once. Each payment cycle will represent a set of rewards for staking (ie set of reward_proofs).

Reply:
Ahh I see you've approved. @jurgenwerk do you know if there is an easy way to test the queue priority

I am not too certain what the expected outcome is in terms of a test when it comes to graphile worker

Reply:
Ok skipping test for this